File: distribution/rip302-go-sdk/client_test.go
package agenteconomy

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestBrowseFilters(t *testing.T) {
	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/agent/jobs" || r.URL.Query().Get("status") != "open" || r.URL.Query().Get("category") != "code" || r.URL.Query().Get("limit") != "5" {
			t.Fatalf("bad request: %s", r.URL.String())
		}
		json.NewEncoder(w).Encode(map[string]any{"jobs": []any{map[string]any{"job_id": "j1"}}})
	}))
	defer s.Close()
	c := New(s.URL, "")
	out, err := c.BrowseJobs(context.Background(), "", "code", 5)
	if err != nil {
		t.Fatal(err)
	}
	if out["jobs"] == nil {
		t.Fatal("jobs missing")
	}
}

func TestWriteWalletBinding(t *testing.T) {
	seen := []map[string]any{}
	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var body map[string]any
		json.NewDecoder(r.Body).Decode(&body)
		seen = append(seen, body)
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"ok":true,"job_id":"j"}`))
	}))
	defer s.Close()
	c := New(s.URL, "wallet-a")
	if _, err := c.PostJob(context.Background(), "x", "code", 5, "", ""); err != nil {
		t.Fatal(err)
	}
	if _, err := c.ClaimJob(context.Background(), "j", ""); err != nil {
		t.Fatal(err)
	}
	if _, err := c.DeliverJob(context.Background(), "j", "", "https://example.invalid", "done"); err != nil {
		t.Fatal(err)
	}
	if seen[0]["poster_wallet"] != "wallet-a" || seen[1]["worker_wallet"] != "wallet-a" || seen[2]["worker_wallet"] != "wallet-a" {
		t.Fatal(seen)
	}
}

func TestAllLifecycleMethods(t *testing.T) {
	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"ok":true}`))
	}))
	defer s.Close()
	c := New(s.URL, "w")
	ctx := context.Background()
	calls := []func() (map[string]any, error){
		func() (map[string]any, error) { return c.GetJob(ctx, "j") },
		func() (map[string]any, error) { return c.AcceptJob(ctx, "j", "", 5, "ok") },
		func() (map[string]any, error) { return c.DisputeJob(ctx, "j", "", "why") },
		func() (map[string]any, error) { return c.CancelJob(ctx, "j", "") },
		func() (map[string]any, error) { return c.Reputation(ctx, "") },
		func() (map[string]any, error) { return c.Stats(ctx) },
	}
	for _, fn := range calls {
		out, err := fn()
		if err != nil || out["ok"] != true {
			t.Fatalf("out=%v err=%v", out, err)
		}
	}
}

func TestHTTPError(t *testing.T) {
	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(503)
		w.Write([]byte(`{"error":"offline"}`))
	}))
	defer s.Close()
	c := New(s.URL, "")
	_, err := c.Stats(context.Background())
	if err == nil {
		t.Fatal("expected error")
	}
	if _, ok := err.(*APIError); !ok {
		t.Fatalf("wrong error %T", err)
	}
}

func TestMissingWalletFailsBeforeNetwork(t *testing.T) {
	c := New("http://127.0.0.1:1", "")
	if _, err := c.ClaimJob(context.Background(), "j", ""); err == nil {
		t.Fatal("expected wallet error")
	}
}

File: distribution/rip302-js-sdk/test/client.test.js
import test from "node:test";
import assert from "node:assert/strict";
import {RustChainAgentEconomy, RustChainAgentEconomyError} from "../src/index.js";

function fakeFetch(routes) {
  return async (url, opts = {}) => {
    const key = `${opts.method || "GET"} ${url.pathname}`;
    const handler = routes[key];
    if (!handler) throw new Error(`unexpected route ${key}`);
    const out = handler(url, opts);
    return {
      ok: out.status >= 200 && out.status < 300,
      status: out.status,
      async text() { return out.text ?? JSON.stringify(out.body ?? {}); },
    };
  };
}

test("browse jobs encodes filters", async () => {
  const client = new RustChainAgentEconomy({
    fetchImpl: fakeFetch({
      "GET /agent/jobs": (url) => {
        assert.equal(url.searchParams.get("status"), "open");
        assert.equal(url.searchParams.get("category"), "code");
        assert.equal(url.searchParams.get("limit"), "5");
        return {status: 200, body: {jobs: [{job_id: "j1"}]}};
      },
    }),
  });
  const out = await client.browseJobs({category: "code", limit: 5});
  assert.equal(out.jobs[0].job_id, "j1");
});

test("post, claim and deliver use configured wallet", async () => {
  const seen = [];
  const client = new RustChainAgentEconomy({
    wallet: "worker-wallet",
    fetchImpl: fakeFetch({
      "POST /agent/jobs": (_u, o) => { seen.push(JSON.parse(o.body)); return {status: 201, body: {job_id: "j1"}}; },
      "POST /agent/jobs/j1/claim": (_u, o) => { seen.push(JSON.parse(o.body)); return {status: 200, body: {ok: true}}; },
      "POST /agent/jobs/j1/deliver": (_u, o) => { seen.push(JSON.parse(o.body)); return {status: 200, body: {ok: true}}; },
    }),
  });
  await client.postJob({title: "Task", category: "code", rewardRtc: 5});
  await client.claimJob("j1");
  await client.deliverJob("j1", {deliverableUrl: "https://example.invalid/result", resultSummary: "done"});
  assert.equal(seen[0].poster_wallet, "worker-wallet");
  assert.equal(seen[0].reward_rtc, 5);
  assert.equal(seen[1].worker_wallet, "worker-wallet");
  assert.equal(seen[2].worker_wallet, "worker-wallet");
});

test("accept, dispute, cancel and reputation are covered", async () => {
  const routes = {};
  for (const [method, path] of [
    ["POST", "/agent/jobs/j1/accept"],
    ["POST", "/agent/jobs/j1/dispute"],
    ["POST", "/agent/jobs/j1/cancel"],
    ["GET", "/agent/reputation/w"],
    ["GET", "/agent/stats"],
    ["GET", "/agent/jobs/j1"],
  ]) routes[`${method} ${path}`] = () => ({status: 200, body: {ok: true}});
  const c = new RustChainAgentEconomy({wallet: "w", fetchImpl: fakeFetch(routes)});
  assert.equal((await c.acceptJob("j1")).ok, true);
  assert.equal((await c.disputeJob("j1", {reason: "bad"})).ok, true);
  assert.equal((await c.cancelJob("j1")).ok, true);
  assert.equal((await c.reputation()).ok, true);
  assert.equal((await c.stats()).ok, true);
  assert.equal((await c.getJob("j1")).ok, true);
});

test("HTTP and JSON errors are explicit", async () => {
  const c1 = new RustChainAgentEconomy({
    fetchImpl: fakeFetch({"GET /agent/stats": () => ({status: 503, body: {error: "offline"}})}),
  });
  await assert.rejects(() => c1.stats(), (e) => e instanceof RustChainAgentEconomyError && e.status === 503);

  const c2 = new RustChainAgentEconomy({
    fetchImpl: fakeFetch({"GET /agent/stats": () => ({status: 200, text: "not-json"})}),
  });
  await assert.rejects(() => c2.stats(), /non-JSON/);
});

test("missing wallet fails before network", async () => {
  const c = new RustChainAgentEconomy({fetchImpl: async () => { throw new Error("must not call network"); }});
  await assert.rejects(() => c.claimJob("j1"), /worker wallet required/);
});
